auditlog/log-enricher/src/main/java/io/debezium/demos/auditing/enricher/TopologyProducer.java

```diff
@@ -0,0 +1,108 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@ApplicationScoped
+public class TopologyProducer {
+
+    private static final String STORE_NAME = "transaction-meta-data";
+
+    private static final Logger LOG = LoggerFactory.getLogger(TopologyProducer.class);
+
+    @ConfigProperty(name = "audit.context.data.topic")
+    String txContextDataTopic;
+
+    @ConfigProperty(name = "audit.vegetables.topic")
+    String vegetablesTopic;
+
+    @ConfigProperty(name = "audit.vegetables.enriched.topic")
+    String vegetablesEnrichedTopic;
+
+    @Produces
+    public Topology buildTopology() {
+        StreamsBuilder builder = new StreamsBuilder();
+
+        builder.globalTable(txContextDataTopic, Materialized.as(STORE_NAME));
+
+        builder.<JsonObject, JsonObject>stream(vegetablesTopic)
+                // filter out any tombstones; for an audit log topic, time-based retention
+                // seems more reasonable than compaction
+                .filter((id, changeEvent) -> changeEvent != null)
+
+                // enrich change events with transaction metadata via the statestore of the TX
+                // topic
+                .transform(() -> new ChangeEventEnricher())
+                .to(vegetablesEnrichedTopic);
+
+        return builder.build();
+    }
+
+    private final class ChangeEventEnricher implements Transformer<JsonObject, JsonObject, KeyValue<JsonObject, JsonObject>> {
+
+        private KeyValueStore<JsonObject, JsonObject> txMetaDataStore;
+
+        @Override
+        @SuppressWarnings("unchecked")
+        public void init(ProcessorContext context) {
+            txMetaDataStore = (KeyValueStore<JsonObject, JsonObject>) context.getStateStore(STORE_NAME);
+        }
+
+        @Override
+        public KeyValue<JsonObject, JsonObject> transform(JsonObject key, JsonObject value) {
+            JsonObject txId = Json.createObjectBuilder()
+                    .add("transaction_id", value.get("source").asJsonObject().getJsonNumber("txId").longValue())
+                    .build();
+
+            JsonObject metaData = getTransactionMetaData(txId);
+
+            if (metaData != null) {
+                value = Json.createObjectBuilder(value).add("audit", metaData).build();
+            }
+
+            return KeyValue.pair(key, value);
+        }
+
+        @Override
+        public void close() {
+        }
+
+        private JsonObject getTransactionMetaData(JsonObject txId) {
+            int i = 1;
+
+            while (i < 1_000) {
+                JsonObject metadata = txMetaDataStore.get(txId);
+
+                if (metadata != null) {
+                    LOG.info("Found TX metadata after {} attempt(s)", i);
+                    return Json.createObjectBuilder(metadata.get("after").asJsonObject())
+                            .remove("transaction_id")
+                            .build();
+                }
+                i++;
+
+                try {
+                    Thread.sleep(1);
```

Comment:
This will cause your consumer to be rebalanced since you are sleeping for > MAX_POLL_MS on the main thread.

Try looking into the `punctuate()` method of the Processor or Transformer API.

Reply:
Hey @hashhar, thanks a lot for giving this a look! Interesting remark with the rebalance; seems there's no way around using `punctuate()` then. What bothers me most about this: I may have to buffer multiple messages with the same key in the stream-side statestore (e.g. two UPDATEs for the same purchase order, whose TX event hasn't arrived yet. Is there any support for that in the state store API (multi-value for a given key), or would I have to collect the events in my own value holder (a JSON array essentially)?

Reply:
@gunnarmorling as far as the state stores are concerned they operate on Bytes(key) and byte[] (value) see e.g. the RocksDB or InMem implementations. This essentially means you either provide De/Serializers for custom data structures (e.g. List of something) or go with JSON Serdes like you suggest.

Reply:
Ok; so this all is getting much more complex than I had anticipated, but it's interesting, too :)

How about the following design/algorithm then:

* have a custom K/V store managed by a stream-side `Processor` ("buffer")
* have a punctuate method on that processor using wall-clock time

* when a stream event (purchase order) arrives
  * when the buffer is not empty
    * process buffered elements in insertion order
      * lookup TX metadata in GlobalKTable store
      * when found -> emit joined record
      * when not found, abort loop
    * handle stream event as in "buffer is empty case"
  * when the buffer is empty
    * lookup TX metadata in GlobalKTable store
      * when found -> emit joined record
      * when not found -> add stream event to buffer
* when punctuate is called
  * when the buffer is not empty: same logic as above

IIUC, there shouldn't be the need for `punctuate()` on the table side of the join, as the wall-clock punctuation on the stream side will process the buffer based on the latest additions to the `GlobalKTable`.

Reply:
@renatomefi, as we were talking about using punctuation before, it seems that's the way to go as per @hashhar's comment above. I remember you mentioned you'd be interested in exploring this route, WDYT about the approach above? Let me know if you're interested, I can also wrap this up (next week likely).